Adding files for the NewsApi challenge

AineKiraboMbabazi opened the pull request:
#NewApi Challenge
-User is offered a list of four news sources.
- User should be able to make a choice of a preferred news source. 
-User gets back a list of the top 10 headlines from the preferred source.
-
![screenshot from 2018-09-17 21-08-41](https://user-images.githubusercontent.com/29704363/45641378-f95ce480-babd-11e8-8290-fa0588353a2e.png)
![screenshot from 2018-09-17 21-09-05](https://user-images.githubusercontent.com/29704363/45641380-fb26a800-babd-11e8-807e-5b5ae8ff180a.png)
![screenshot from 2018-09-18 15-06-22](https://user-images.githubusercontent.com/29704363/45686912-01686300-bb56-11e8-9d85-ff0f87a0085b.png)

 A news headline should have a title, description and a url in case the user needs to follow up on the news story.

AineKiraboMbabazi committed:
Adding files for the NewsApi challenge

AineKiraboMbabazi commented:
@dennisja 
@janetnamutebi
@malmike
please Review my solution for the newsapi challenge

AineKiraboMbabazi committed:
newsapi.py ../tests/test_newsapi.py
Adding a table to display the headlines to the newsapi script.newsapi.py

dennisja commented:
Remove all files in the `.pytest_cache` folder from version control

dennisja commented on apicode/newapi.py:
```diff
@@ -0,0 +1,76 @@
+        print("\t", '{: ^100}'.format('Top 10 Headlines from '))
+        print("\t", '{:=^100}'.format('.'))
+        for article in content:
+            print("\t", '{:_^100}'.format(article_count))
```
Would appreciate if you could just build a string and print it out using one print function

AineKiraboMbabazi replied:
I have addressed this issue by using a texttable module in python

dennisja commented on apicode/newapi.py:
```diff
@@ -0,0 +1,76 @@
+        if('API_KEY' in os.environ) == False:
+            raise ValueError(
+                "Access Denied ..., you dont have a product key in your environment variables")
+        return os.environ.get('API_KEY')
```
Good attempt of getting the apiKey from environment variables

AineKiraboMbabazi replied:
thank you, it was Malmikes idea.

dennisja commented on apicode/newapi.py:
```diff
@@ -0,0 +1,76 @@
+        return answers['sources']
+    # function to retrieve a product key for the environment variable
+
+    def add_api_key(self):
```
This method name is annoying. It would be better if you could name it `get_api_key`

AineKiraboMbabazi replied:
Hahaha to be changed immediately

dennisja commented on apicode/newapi.py:
```diff
@@ -0,0 +1,76 @@
+
+    def fetch_data(self, newssource):
+
+        api_key = self.add_api_key()
```
The api key method raises an exception.
When using such a method, you should wrap it in a `try...except` block to catch any errors that might occur

dennisja replied:
This helps prevent your app from crashing

dennisja commented on apicode/newapi.py:
```diff
@@ -0,0 +1,76 @@
+    def sort_and_display_results(self):
+        content = self.store_api_response()
+        article_count = 1
+        print('articles length', len(content))
```
Build a string and print it using 1 print statement

dennisja commented on apicode/newapi.py:
```diff
@@ -0,0 +1,76 @@
+
+        return headlineArticles
+
+    def sort_and_display_results(self):
```
This method is called `sort_and_display_results` but, I don't see any where you are sorting the results from

AineKiraboMbabazi committed:
Addressing concerns in the previous version

dennisja commented on apicode/newapi.py:
```diff
@@ -0,0 +1,100 @@
+        article_count = 1
+        source_selected = content[1]
+
+        print('Headlines retrieved : ', len(content[0]))
```
Build this using a single string

dennisja commented on apicode/newapi.py:
```diff
@@ -0,0 +1,100 @@
+    # function to retrieve a product key for the environment variable
+
+    def get_api_key(self):
+        try:
```
This is unneccessary. The code in except will never be executed, The get method returns none if no key is found

dennisja commented on apicode/newapi.py:
```diff
@@ -0,0 +1,100 @@
+        except Exception:
+
+            if 'API_KEY' not in os.environ:
+                print(
```
Would be better to raise an exception, to show that the api key is missing

dennisja commented:
Include the dependencies of your application in a `requirements` file
Update your readme to indicate how a person should set up your application and run tests and use it

AineKiraboMbabazi committed:
Adding fixes to the code
to address issues in the pull request
adding requirements.txt file
adding modifications to the readme file

dennisja commented on README.md:
```diff
@@ -1,2 +1,60 @@
 # NewsAPI
 User is offered a list of four news sources. User should be able to make a choice of a preferred news source. User gets back a list of the top 10 headlines from the preferred source. A news headline should have a title, description and a url in case the user needs to follow up on the news story.
+
+# Requirements
```
These shouldn't be ~~packages~~ dependencies cuz they are there in the requirements file. These should be the things one should have inorder to setup your project

AineKiraboMbabazi replied:
I am having trouble understanding this comment, did you mean that i should put things like the user should have python 3.6 installed?

dennisja replied:
You are specifying your project dependencies yet they are already in the requirements.txt file. The requirements here should be things one should have set up to run your project e.g One should have PythonXX, e.t.c

dennisja commented on README.md:
```diff
@@ -1,2 +1,60 @@
+# Installation and settup
+Clone the repository
+
+$ git clone https://github.com/AineKiraboMbabazi/NewsAPI.git
```
Please make this a code block. 
Make all other commands code blocks

dennisja replied:
Read this [guide to write a better readme](https://guides.github.com/features/mastering-markdown/)

dennisja commented on apicode/newapi.py:
```diff
@@ -0,0 +1,99 @@
+        numberofitems = str(len(content[0]))
+        title = "TOP 10 HEADLINES FROM : "
+        #creating the header for the display table.
+        print('%s %s \n %s \n \t\t\t %s %s \n %s' % ('Headlines retrieved : ',
```
Using `.format` is better. You can look into how you can use it [here](https://pyformat.info/)

And if you are using python 3.6+
`f strings` are the best. You can look at them [here](https://www.python.org/dev/peps/pep-0498/?)

dennisja commented on apicode/newapi.py:
```diff
@@ -0,0 +1,99 @@
+            '&apiKey='+api_key+'&pageSize=10'
+
+        return (requests.get(url))
+    # function for checking whether the api call was successfull
```
Make these comments doc strings and put them at the start of the function e.g.
```python
# good
def my_awesome_method():
     """ Does awesome things """
     do_something_awesome()

# bad
# does awesome things
def my_awesome_method():
   do_something_awesome()
```

AineKiraboMbabazi replied:
noted

AineKiraboMbabazi committed:
Adding code blocks to the readme
Add the .format to the newapi file

AineKiraboMbabazi committed:
add .travis.yml file to the repository
.

AineKiraboMbabazi committed:
add pytest to the requirements file

dennisja commented on .travis.yml:
```diff
@@ -0,0 +1,11 @@
+  - pip install -r requirements.txt
+# command to run tests
+script:
+  - pytest
```
Point pytest to the folder containing your application ~~code~~ tests to fix the Travis build
```yml
script:
  - pytest tests
```

dennisja replied:
@AineKiraboMbabazi 👆 I have updated this. Sorry for the initial info. I was on rush. The first comment applies to coverage

AineKiraboMbabazi committed:
add doc format comments in the newapi.py file
add apicode to fix build error

AineKiraboMbabazi committed:
add unittest

AineKiraboMbabazi committed:
remove unittest

AineKiraboMbabazi committed:
adding -s to the pytest command

AineKiraboMbabazi committed:
add new command for running tests

AineKiraboMbabazi committed:
Add mock for the test get_user_source function inthe test file

AineKiraboMbabazi committed:
add mock for the get api function

AineKiraboMbabazi committed:
add api from the environment variable

AineKiraboMbabazi committed:
add travis-ci badge

dennisja commented:
Good Job on running tests.
Can you do the same on coverage?
look at how you can leverage https://pypi.org/project/pytest-cov/ and coveralls for test reporting

AineKiraboMbabazi commented:
Thank you.
Am getting started on it right now

AineKiraboMbabazi committed:
add coverage and coveralls

AineKiraboMbabazi committed:
add coveralls

dennisja commented on .coveralls.yml:
```diff
@@ -0,0 +1 @@
+repo_token: zbmrHOitciQcFigB3CwnHTpTpqp1gn0lV
```
What is the repo token doing here?

dennisja commented on .coverage:
```diff
@@ -0,0 +1 @@
+!coverage.py: This is a private format, don't read it directly!{"lines":{"/home/andela/Projects/cli_tests/app/__init__.py":[1],"/home/andela/Projects/cli_tests/app/app_Api.py":[1,2,3,4,5,8,9,11,28,36,49,61,70,106,63,38,41,42,30,31,34,45,47,64,66,68,52,53,54,55,56,58,73,74,75,76,77,78,80,81,84,85,86,87,89,90,91,92,93,94,95,96,97,98,99,101,102,103]}}
```
The .coverage file shouldn't be in version control

dennisja commented on .travis.yml:
```diff
@@ -0,0 +1,17 @@
+  
+# command to run tests
+script: 
+  - pytest tests 
```
This command doesn't get coverage

AineKiraboMbabazi committed:
add coveralls
add coverage to gitignore

AineKiraboMbabazi committed:
add the repo token

dennisja commented on .travis.yml:
```diff
@@ -6,12 +6,12 @@ python:
 install:
   - pip install -r requirements.txt
-  - pip install coverage
+  - pip install pytest-cov
```
This should be in your requirements.txt

dennisja commented on .travis.yml:
```diff
@@ -6,12 +6,12 @@ python:
 script: 
-  - pytest tests 
+  - pytest --cov=app/
   - coveralls 
```
You are running this command twice.
Here and in the after scripts block

AineKiraboMbabazi committed:
edit travis.yml

AineKiraboMbabazi committed:
edit travis.yml
